xds/internal/xdsclient/transport/ads/ads_stream.go

```diff
@@ -0,0 +1,780 @@
+type Stream struct {
+	// The following fields are initialized from arguments passed to the
+	// constructor and are read-only afterwards, and hence can be accessed
+	// without a mutex.
+	transport          transport.Interface     // Transport to use for ADS stream.
+	eventHandler       StreamEventHandler      // Callbacks into the xdsChannel.
+	backoff            func(int) time.Duration // Backoff after stream failures.
+	nodeProto          *v3corepb.Node          // Identifies the gRPC application.
+	watchExpiryTimeout time.Duration           // Resource watch expiry timeout
+	logger             *igrpclog.PrefixLogger
+
+	// The following fields are initialized in the constructor and are not
+	// written to afterwards, and hence can be accessed without a mutex.
+	streamCh     chan transport.StreamingCall // New ADS streams are pushed here.
+	requestCh    *buffer.Unbounded            // Subscriptions and unsubscriptions are pushed here.
+	runnerDoneCh chan struct{}                // To notify exit of runner goroutine.
+	cancel       context.CancelFunc           // To cancel the context passed to the runner goroutine.
+
+	// Guards access to the below fields.
+	mu                sync.Mutex
+	resourceTypeState map[xdsresource.Type]*resourceTypeState // Map of resource types to their state.
+	fc                *adsFlowControl                         // Flow control for ADS stream.
+	firstRequest      bool                                    // False after the first request is sent out.
+}
+
+// StreamOpts contains the options for creating a new ADS Stream.
+type StreamOpts struct {
+	Transport          transport.Interface     // xDS transport to create the stream on.
+	EventHandler       StreamEventHandler      // Callbacks for stream events.
+	Backoff            func(int) time.Duration // Backoff after stream failures.
+	NodeProto          *v3corepb.Node          // Node proto to identify the gRPC application.
+	WatchExpiryTimeout time.Duration           // Resource watch expiry timeout.
+	LogPrefix          string                  // Prefix to be used for log messages.
+}
+
+// NewStream initializes a new ADS Stream instance using the given parameters.
+// It also launches goroutines responsible for managing reads and writes of
+// messages with the underlying stream.
+func NewStream(opts StreamOpts) *Stream {
+	s := &Stream{
+		transport:          opts.Transport,
+		eventHandler:       opts.EventHandler,
+		backoff:            opts.Backoff,
+		nodeProto:          opts.NodeProto,
+		watchExpiryTimeout: opts.WatchExpiryTimeout,
+
+		streamCh:          make(chan transport.StreamingCall, 1),
+		requestCh:         buffer.NewUnbounded(),
+		runnerDoneCh:      make(chan struct{}),
+		resourceTypeState: make(map[xdsresource.Type]*resourceTypeState),
+	}
+
+	l := grpclog.Component("xds")
+	s.logger = igrpclog.NewPrefixLogger(l, opts.LogPrefix+fmt.Sprintf("[ads-stream %p] ", s))
+
+	ctx, cancel := context.WithCancel(context.Background())
```

Comment:
Yeah after reading this the synchronization between layers/the incorrect scoping of the lrsStreams reference count going to 0. I see how you need a context here to cancel all the operations spawned by this component, I just wish there was another signaling mechanism we could use instead. Another thing I was thinking is at least allowing context to be configurable since the higher layer does technically create this/start operations so it theoretically could own a linked context, but the good thing is it can also manage this components lifecycle and Close cleans up spawned resources fine. The other reason I brought this up is due to GOA conformance tests I "broke", I know it's considered against best practices to

1. Call context.Background() from somewhere not in main and
2. Persist a cancel func or a context in a component.

Thinking about it I guess xDS Clients are not "underneath" a gRPC Channel or Server in the same manner a LB policy is, so the top level channel/server context provided from main() is not scoped correctly, and the argument at a higher level within the xDS Client is the points you provided above. Also this is logically a stream, so like how application main() have a context passed into RPC's something needs to provide a context in this case of creating a stream. So I'm fine leaving this, but I'd like to hear a technical rebuttal/explanation as to why we're ok going against best practices here.

Reply:
I think the important distinction here is that most of what is happening in the `ads.Stream` and the `lrs.Stream` are not in response to any single API call chain. They are essentially long running asynchronous events. Using a `context` for cancelation in this scenario seems not the best use for it. Instead a `Stop` method like what we have today is better off, since `Stop` can actually block until everything that the `ads.Stream` spawned has been closed.

Reply:
> so the top level channel/server context provided from main() is not scoped correctly

There is no top-level context for a grpc channel or server.

Reply:
This is what I mean when I say "top level context" for a channel, it's not passed in by the user of the API but it does have a "top level context" and cancel func https://github.com/grpc/grpc-go/blob/master/clientconn.go#L140. Oh yeah, the server doesn't get an explicit context I guess, just relies on Stop() which waits for spawned goroutines to finish.

The way I see it, the issue is how to think of context scoping. For OpenTelemetry, we wanted it to be configurable because OpenTelemetry operations take a context: https://github.com/grpc/grpc-go/blob/master/stats/opentelemetry/opentelemetry.go#L335, but after triaging the code we realized that this the context is not used and is a no-op, so we just decided to add a context.TODO. The reason we viewed it as configurable is the scope of OpenTelemetry instrumentation is logically distinct from the Channel/Server, and so we wanted to provide scoping to an applications main(), as the context/scoping are logically distinct.

In the case of these hierarchy of components, it seems like the scoping is distinct across layers, (i.e. channels don't wrap an xDS Client 1:1, and we now share different components along this hierarchy). However, the base layer uses a spawned context in it's synchronization (i.e. it's a stream, so it passes the context to operations), and creating a context outside of main goes against Go style, so I don't know how to reconcile that. In the case for Google production, I think a context should be passed to these streaming operations as an implementor can now get a signal from a timeout, but I don't know what layer that this context should live at, as the hierarchy/layering isn't composable and thus I don't know how to scope the context.Cancel(). Perhaps just leaving it as is and Close() on these stream types having cleanup functions which propagate cancellation signal is fine, just like the xDS Server.

@dfawley thoughts on this? I know you've had issues with the way we persist context around in ClientConn/other components in the past.